### ISSUE TITLE
[clarification]: list of restrictions on repository method signatures in Javadoc of PageRequest

### ISSUE THREAD

**Issue by gavinking** [jakartaee/data#518]
### Specification

PageRequest

### I need clarification on ...

should those restrictions be stated elsewhere.

### Additional information

In the Javadoc for `PageRequest`, we have this text:

> A repository method will fail with a DataException or a more specific subclass if
>
> - multiple PageRequest parameters are supplied to the same method.
> - PageRequest and Limit parameters are supplied to the same method.
> - a PageRequest parameter is supplied in combination with the First keyword.
> - a PageRequest parameter with sort criteria is supplied and separate Sort parameters are also supplied to the same method.
> - a PageRequest parameter with sort criteria is supplied and an Order parameter is also supplied to the same method.
the database is incapable of ordering with the requested sort criteria.

This seems to mix together two sorts of disallowed things:

- The first three conditions are things which ideally&mdash;at least in an implementation based on annotation processing&mdash;would be rejected at compile time. They form part of the programming model for repository methods.
- The last two conditions seem to be things which are only usually detectable at runtime.

Also, I wonder why this is always a `DataException`, whereas in spec section 4.5, a similar sort of problem (to the first three cases considered here) is specified to result in `UnsupportedOperationException`.

I'm thinking that this information doesn't belong on `PageRequest`, and should probably be moved to section 4.7 of the specification, which should also mention the possibility of rejecting such method signatures at compile time.

**Comment by njr-11** [comment]
> Also, I wonder why this is always a `DataException`, whereas in spec section 4.5, a similar sort of problem (to the first three cases considered here) is specified to result in `UnsupportedOperationException`.

It can be difficult to decide which exception class to use sometimes. `IllegalArgumentException` would have been another possibility.  I'm fine with `UnsupportedOperationException` if other see that fitting better here, and also for consistency with elsewhere in the spec as pointed out.

> I'm thinking that this information doesn't belong on `PageRequest`, and should probably be moved to section 4.7 of the specification, which should also mention the possibility of rejecting such method signatures at compile time.

I'm fine with including it in the specification document, but I wouldn't remove it from the JavaDoc because the JavaDoc will help prevent users from attempting these error paths. Users are much more likely to read a method's JavaDoc before using the method than they are to read the spec before using a method.

**Comment by gavinking** [comment]
> I wouldn't remove it from the JavaDoc because the JavaDoc will help prevent users from attempting these error paths

Yeah I'm not even saying remove it from the javadoc entirely, more like suggesting that this is not the right place for it. I would put that info in the jdoc for `@Find` and `@Query`, since it's part of the contract for those annotations, not really part of the contract for `PageRequest`.

**Comment by gavinking** [comment]
> It can be difficult to decide which exception class to use sometimes. 

I don't really care _which_ exception, just that it be consistent.

**Comment by njr-11** [comment]
> Yeah I'm not even saying remove it from the javadoc entirely, more like suggesting that this is not the right place for it. I would put that info in the jdoc for `@Find` and `@Query`, since it's part of the contract for those annotations, not really part of the contract for `PageRequest`.

`Find` and `Query` won't cover all usage because you can also use it with Query-by-Method-Name, which has no annotation or specific API class JavaDoc of its own.  I think that's why this language is on `PageRequest` and `Limit` and `Sort` and `Order`.

**Comment by gavinking** [comment]
I have sent https://github.com/jakartaee/data/pull/530 which would solve this issue from my point of view.

**Comment by gavinking** [comment]
@njr-11 please note:

> This seems to mix together two sorts of disallowed things:
>
> - The first three conditions are things which ideally—at least in an implementation based on annotation processing—would be rejected at compile time. They form part of the programming model for repository methods.
> - The last two conditions seem to be things which are only usually detectable at runtime.

**Comment by gavinking** [comment]
This was fixed a while ago.
